Draft of Notifications: Trash e2e fix utilizing REST API

john-legg opened the pull request:
### Context
The purpose of this PR is to make the Notifications: Trash e2e test more reliable [#66144](https://github.com/Automattic/wp-calypso/issues/66144). At some point, this test was passing inconsistently and was tagged as quarantined. Although it passes more consistently now, the test could benefit from some changes, such as using the Wordpress.com REST API to set up the notification used for the test

### Completed Steps
- [x] Create a comment using the REST API to generate a notification for the "notification user."
- [x] Add additional actions that a user can perform on a notification (approve, like, mark as spam).
- [x] Add undo actions for each action above (e.g. un-approve).
- [x] Rename the test to better reflect all notification flows, not just trash.
- [x] Add a clean up method that deletes the test comment using the REST API. 
- [x] Un-quarantine the test. 

### Future Steps (Outside of Scope)
- [ ] Create the test post used for commenting if it's not found.
- [ ] Handle the case where there are multiple posts (right now it's just one test post).

john-legg committed:
Draft of Notifications: Trash e2e fix utilizing REST API

worldomonation commented:
I've converted this PR to a draft. for now.

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -166,6 +166,68 @@ export class RestAPIClient {
+	 * Gets the latest of post from the blogs a user follows.
+	 *
+	 *
+	 * @returns {Promise<ReaderParams}.
```
**nit**:the type declaration is missing the closing angular bracket.

john-legg replied:
Good catch! Just fixed it.

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -166,6 +166,68 @@ export class RestAPIClient {
+			},
+		};
+
+		const response = await this.sendRequest( this.getRequestURL( '1.1', '/me/sites' ), params );
```
**comment**: the `/me/sites` endpoint return an array of site data, represented by the interface `AllSitesResponse`, not `ReaderParams`.

This may be one of the reasons why the code doesn't work currently.

john-legg replied:
I have no idea how `/me/sites` got in there instead of `read/following` 😅 It's all fixed now though!

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -166,6 +166,68 @@ export class RestAPIClient {
+		siteID: number,
+		postID: number,
+		comment: string
+	): Promise< NewCommentParams > {
```
**comment**: this specifies the return type for this method is an implementation of the `NewCommentParams` interface, but I checked the REST API and it instead returns the following:

```
{
  "found": 0,
  "site_ID": 1,
  "comments": []
}
```

worldomonation reviewed (commented):
Overall, a good first revision. I noted some data type/interface issues that may be the cause of your issues.

worldomonation commented on packages/calypso-e2e/src/types/rest-api-client.types.ts:
```diff
@@ -26,6 +26,16 @@ export interface NewPostParams {
 	content?: string;
 }
 
+export interface ReaderParams {
+	postID: number;
+	siteID: number;
+}
+
+export interface NewCommentParams {
+	ID: number;
+	content: string;
+}
```
**comment**: typically when a new REST API method is implemented in `RestAPIClient`, you'd need to create both the Params and Response versions - just a heads up!

john-legg replied:
Ah, thanks for the heads up! I understand it a lot better now 👍

john-legg committed:
Fixing REST API flow so it actually works :)

john-legg commented:
### Update
I have the REST API flow working now and I've updated the description above to reflect the next steps. @worldomonation and @dpasque I have a few questions below - more just out of curiosity.

**question: non-blocking**

Is there a way to do filtering on WordPress.com API calls? For example, I'd like to only get posts from user `e2eflowtestingnotifications` on the `e2eflowtestingnotifications` site using the [reader feed endpoint](https://developer.wordpress.com/docs/api/1.1/get/read/following/). I don't see any [query parameters](https://developer.wordpress.com/docs/api/1.1/get/read/following/#apidoc-query) that could help. Right now, the endpoint returns only one, which is the one I need, but that might not always be the case in the future. 

**question: non-blocking**
Is there an elegant way to include query parameters (in the URL, not body) with REST API calls in `rest-api-client.ts`? For example, the [reader feed endpoint](https://developer.wordpress.com/docs/api/1.1/get/read/following/) has a [parameter](https://developer.wordpress.com/docs/api/1.1/get/read/following/#apidoc-query) to limit the number of posts returned. I could add the query on at the end of the URL (shown below), but I wasn't sure if there was a cleaner method 😄 

```
const response = await this.sendRequest(
	this.getRequestURL( '1.1', '/read/following' + '?number=1' ),
	params
);
```

john-legg committed:
Adding additional action flows for notifications

john-legg commented:
@dpasque thanks answering my questions! I will definitely use those tips for my next update 🙇  @worldomonation, here's a quick summary of the latest update (no rush by the way - I know you're still catching up on things).

### Update
I just added additional flows available to users when viewing notifications:
- Approve/un-approve
- Like/un-like
- Mark as spam/undo mark as spam
- Trash/undo trash

I'll handle the `edit` action in another update. Also, now that there are additional flows, I've renamed the spec from `notifications__trash.ts` to `notification__actions.ts`. If there's a more appropriate name that I should use, please let me know! 

**question: non-blocking**
When logged in as the `e2eflowtestingnotifications` user, it appears that the `edit` button on the notification allows me to edit `e2eflowtestingcommenter`'s comment. Is that an expected functionality? (To be able to edit someone else's comment on your site and post).

### Next Steps
- [ ] Add `edit` flow.
- [ ] Generate a post (to comment on) if the test post is not there.
- [ ] Handle case where there are multiple posts on the site (filter and parse after response).

worldomonation commented:
> When logged in as the e2eflowtestingnotifications user, it appears that the edit button on the notification allows me to edit e2eflowtestingcommenter's comment. Is that an expected functionality? (To be able to edit someone else's comment on your site and post).

@john-legg Yes, this functionality is expected. 

You can also try it out for yourself with a personal WordPress.com account, though because we have superuser privileges we are able to edit more than normal users. 

> I've renamed the spec from notifications__trash.ts to notification__actions.ts

I think the new name is good 👍

john-legg commented:
I had a great sync session with @worldomonation today. We determined that the edit flow is not necessary for the scope of this PR. In addition, generating a new post (to comment on) if the test post is not there and handling the case where there are multiple posts on the site would be "nice-to-haves," and therefore outside of this scope as well.

**tl;dr:** I will clean up the test in its current state, rebase (scary), and push it tomorrow.

john-legg committed:
Cleaning up the test and adding an afterAll method to delete test comment.

john-legg committed:
Last minute cleanup

john-legg commented:
@worldomonation looks like the rebase was successful! I cleaned up the test files and included a new `afterAll` method to delete the test comment afterward. I'll update the description with some of my notes for now, but please let me know if you have any questions or suggestions 😄

john-legg commented:
@worldomonation I also had a few quirks that I ran into while finalizing the test that I wanted to mention in case it needs more attention!

**issue: non-blocking**
For the approve/un-approve scenario, I toggled the setting you suggested: `Settings > Discussion > Comment must be manually approved` and it worked great! However, after a few hours my test started to fail again. I checked the setting and it was toggled off again! I'm not sure if it's deployment-related or shared user-related, so I updated my test to just click on the button that contains the word `Approve`, which would handle if it's already approved (`Approved`) or not yet approved (`Approve`). It's not elegant, but it works every time 😅 

**issue: non-blocking**
With the REST API, I noticed that the "commenting user" is not able to use the [delete endpoint](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/comments/%24comment_ID/delete/) to delete its own comment. I have to authenticate with the "notifications user" to delete it. It _is_ the notification user's site and post (I believe), but it seems strange. It could just be a permission/role thing with the API token.

**thought: security**
I accidentally used the API to get a bearer token with an incorrect username and got the following response. I don't remember which security training I learned this from, but I recall hearing that it's not recommended to disclose which part of a login is incorrect (username or password) since it gives attackers more insight. I didn't look too much into it, but that was the first thing that popped in my head 😆 
```
{
    "success": false,
    "data": {
        "errors": [
            {
                "code": "invalid_username",
                "message": "We don't seem to have an account with that name. Double-check the spelling and try again!"
            }
        ]
    }
}
```

worldomonation commented:
> I checked the setting and it was toggled off again! I'm not sure if it's deployment-related or shared user-related, so I updated my test to just click on the button that contains the word Approve

Hmm, this is odd. I would first check the blogRC and check the events log. I did that and didn't see anything too peculiar, so I'm not sure why that setting reverted. 

I'd try flipping that setting back on and assume it will stay that way when constructing your tests. We can always create a new user to replace the existing user to start from a clean slate, if required.

> I have to authenticate with the "notifications user" to delete it. It is the notification user's site and post (I believe), but it seems strange.

I think this is behaving as expected. The notificationUser is the one receiving the notification - thus is the owner of the post, and any comments that gets posted there. While the commentingUser is able to delete the comments, the likely decision was to assign ownership of the comment to the post's owner.

> I recall hearing that it's not recommended to disclose which part of a login is incorrect (username or password) since it gives attackers more insight.

Yep, you're correct - but this would require a change in the backend REST API code, which would require a sandbox. Maybe a great way for you to dip your feet into backend work? 😛

worldomonation commented on packages/calypso-e2e/src/lib/components/notifications-component.ts:
```diff
@@ -64,6 +75,7 @@ export class NotificationsComponent {
 	 * @returns {Promise<void>} No return value.
 	 */
 	async waitForUndoMessageToDisappear(): Promise< void > {
+		this.waitForUndoMessage();
```
**comment:blocking**: FYI, this will not behave in the way you expect, since you're missing the `await` keyword.

worldomonation replied:
There's another occurrence of this higher up in the file too.

john-legg replied:
Great catch, Edwin! Just updated it.

worldomonation commented on test/e2e/specs/infrastructure/notification__actions.ts:
```diff
@@ -0,0 +1,127 @@
+	let siteID: number;
+	let commentID: number;
+
+	describe( `Leave a comment as ${ commentingUser }`, function () {
+		let postID: number;
+
+		beforeAll( async function () {
+			restAPIClient = new RestAPIClient( SecretsManager.secrets.testAccounts.commentingUser );
+		} );
+
+		it( 'Get the latest post in reader stream', async function () {
+			const response: ReaderResponse = await restAPIClient.getReaderFeed();
+			siteID = response.posts[ 0 ].site_ID;
+			postID = response.posts[ 0 ].ID;
+		} );
+
+		it( 'Leave a comment on the post', async function () {
+			const response: NewCommentResponse = await restAPIClient.createComment(
+				siteID,
+				postID,
+				comment
+			);
+			commentID = response.ID;
+		} );
+	} );
```
**style:non-blocking**: The way you have laid out the setup process is perfectly fine. 

For other specs in the repo, the tendency is to have a `beforeAll` at the top level that would perform the required setup, only assigning `describe` blocks and `it` steps to actual test steps.

eg.

```
describe( Some test suite )
    beforeAll()
         log in
         leave a comment on the post
    
    describe ( User is able to see comment )
         it ( test step 1...)
```

Again, there's nothing wrong with giving the test setup its own `describe` and `it` steps, though at the expense of adding to the step count reported in TeamCity and Allure.

john-legg replied:
Makes sense! I'd rather be more in line with the other specs, so I updated the test to have a top level `beforeAll` like you suggested. Thanks!

worldomonation commented on test/e2e/specs/infrastructure/notification__actions.ts:
```diff
@@ -0,0 +1,127 @@
+	describe( `View notification as ${ notificationsUser }`, function () {
+		let page: Page;
+
+		beforeAll( async function () {
+			page = await browser.newPage();
+			const testAccount = new TestAccount( notificationsUser );
+			await testAccount.authenticate( page );
+		} );
```
**style:non-blocking**: I'd say this should be moved out to the top level `beforeAll`, since the setup contained here isn't particular to the `View notification` block.

No harm leaving it here either, hence it's just a non-blocking comment.

john-legg replied:
Sounds good. I've moved it to the top level one in the new commit. Thank you!

worldomonation commented on test/e2e/specs/infrastructure/notification__actions.ts:
```diff
@@ -0,0 +1,127 @@
+	} );
+
+	// Delete test comment that was created.
+	afterAll( async function () {
```
**style:non-blocking**: Similar to above, I'd say this should be moved out to the top level `afterAll` for styling consistency between specs.

john-legg replied:
@worldomonation Thanks for the suggestion! Do you mind clarifying this a bit, please? The current `afterAll` is under the `describe( Some test suite )` level. Is there a level higher than that or do you mean moving it to a higher line position in the file (still under the test suite `describe`)?

worldomonation replied:
Ah is it already at the level within the top level `describe`? The indentation made it seem like it was nested inside a second-level `describe`. Sorry about that, ignore the comment in that case.

john-legg replied:
No worries! I just wanted to double check. Thanks for going through everything.

worldomonation reviewed (changes requested):
Overall, I found this PR to be in a very good state and almost ready to go. As noted, a couple of minor blocking issues that will impact how the script runs, and a few style comments.

Don't take the style comment personally, as the Style Guide hasn't been updated in half a year and it doesn't contain any of the guidances I left in the comments. I'll rectify that issue by updating the documentation soon.

I'll mark this as "Request Changes" but only because of the blocking issues.

john-legg committed:
Fixing code style and missing await keyword

john-legg commented:
> I would first check the blogRC and check the events log.

Ah, great idea! I didn't think to use that! I looked at the audit trail, but I didn't see anything related to the setting. I also don't see any mention of me toggling the setting back on though 😓 I'll check back later and see if it gets turned off again!

> I think [authenticating with notificationUser to delete commentingUser's comment] is behaving as expected.

Thanks for confirming!

> Yep, you're correct - but this would require a change in the backend REST API code, which would require a sandbox. Maybe a great way for you to dip your feet into backend work? 😛

Sure! I'll take note of it and make a GitHub issue to work on later (perhaps after I finish the other quarantine test cleanup ticket).

> Don't take the style comment personally, as the Style Guide hasn't been updated in half a year and it doesn't contain any of the guidances I left in the comments.

Oh no, that was perfect, actually! You have a much better idea of what the consistent style is across the e2e test suite, so I really appreciate you pointing it out. It'll make writing future tests faster 😄 

**issue/question: non-blocking**

In regards to the `E2E Tests (mobile) (Web app)` build failing, I did some reruns in TeamCity and debugging locally and it appears that the following pop up was causing a timeout before the test opened the notification panel:
<img width="410" alt="Top Navigation Pop Up" src="https://user-images.githubusercontent.com/96805492/200080213-4bec0f13-bb4e-4f56-aa33-6f55fa35b7d6.png">
After running the test a few more times, the pop up never returned. I even reran it again in TeamCity and it passed. I'm guessing it's a limited time pop up (maybe after an update or something)? Have you run into that before?

@worldomonation thanks again for thoroughly reviewing my code. I just pushed my latest commit, which should address all of the suggestions you had. If you see anything else that should be looked at, please let me know!

worldomonation commented:
@john-legg 

> In regards to the E2E Tests (mobile) (Web app) build failing, I did some reruns in TeamCity and debugging locally and it appears that the following pop up was causing a timeout before the test opened the notification panel:

Interesting, as you say it might have been one of those temporarily things. From time to time, existing and new accounts will require manual intervention when these "one-time" events occur.

FYI, if you open TeamCity, select a build, then click on the **Artifacts** tab, it might be a faster way to debug than running it locally.

worldomonation reviewed (approved):
Approving for merge.

john-legg commented:
I've noticed the test has been failing intermittently on other branches, specifically the `View notification as notificationsUser` step of the test. It appears to be timing out, but I'll need to investigate further.
